pkg/chart/repo/repo.go

```diff
@@ -0,0 +1,284 @@
+package repo
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"path"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	// Importing this yaml package is a very crucial point:
+	// the "classical" yaml.v2 does not understand json annotations
+	// in structure definitions and therefore always parses empty
+	// index structures. This version is patched to understand json
+	// annotations and works fine.
+	"github.com/Masterminds/semver"
+	yaml "github.com/ghodss/yaml"
+
+	"k8s.io/helm/pkg/chartutil"
+	"k8s.io/helm/pkg/proto/hapi/chart"
+	"k8s.io/helm/pkg/repo"
+
+	"github.com/bookingcom/shipper/pkg/metrics/instrumentedclient"
+)
+
+var (
+	ErrInvalidConstraint = errors.New("invalid constraint")
+	ErrNoneMatching      = errors.New("no matching version found")
+)
+
+type Repo struct {
+	url   string
+	cache Cache
+	mutex sync.Mutex
+}
+
+func NewRepo(repoURL string, cache Cache) *Repo {
+	return &Repo{url: repoURL, cache: cache}
+}
+
+const (
+	RefreshIndexTimeout = 5 * time.Second
+)
+
+func (r *Repo) RefreshIndex() error {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+
+	parsed, err := url.ParseRequestURI(r.url)
+	if err != nil {
+		return fmt.Errorf("failed to parse repo URL: %v", err)
+	}
+	parsed.Path = path.Join(parsed.Path, "index.yaml")
+	indexURL := parsed.String()
+
+	data, err := doFetch(indexURL)
+	if err != nil {
+		return fmt.Errorf("failed to fetch %q: %v", indexURL, err)
+	}
+
+	if _, err := loadIndex(data); err != nil {
+		return fmt.Errorf("failed to load index file: %v", err)
+	}
+
+	if err := r.cache.Store("index.yaml", data); err != nil {
+		return fmt.Errorf("failed to cache index.yaml: %v", err)
+	}
+
+	return nil
+}
+
+func (r *Repo) ResolveVersion(chart, version string) (*repo.ChartVersion, error) {
+	versions, err := r.ChartVersions(chart, version)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get chart versions: %s", err)
+	}
+
+	if len(versions) == 0 {
+		return nil, repo.ErrNoChartVersion
+	}
+
+	var highestver *repo.ChartVersion
+	var lasterr error
+	for _, ver := range versions {
+		if _, lasterr = r.Fetch(ver); lasterr == nil {
```

Comment:
This sounds to me as always trying to fetch charts and not really using the cache a whole lot. Is that on purpose? If not, I would've expected to see `FetchIfNotCached` here.